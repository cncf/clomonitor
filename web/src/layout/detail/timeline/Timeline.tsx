import classNames from 'classnames';
import { isUndefined, uniq } from 'lodash';
import moment from 'moment';
import { Dispatch, Fragment, SetStateAction, useEffect, useState } from 'react';
import { RiHistoryFill } from 'react-icons/ri';

import { SortedDates } from '../../../types';
import groupDatesByYearAndMonth from '../../../utils/groupDatesByYearAndMonth';
import styles from './Timeline.module.css';

interface Props {
  snapshots: string[];
  activeDate?: string;
  currentSearch?: string;
  setActiveDate: Dispatch<SetStateAction<string | undefined>>;
}

const MIN_NUMBER_SNAPSHOTS = 3;

const Timeline = (props: Props) => {
  const [today] = useState<string>(moment().format('YYYY-MM-DD'));
  const [availablaSnapshots, setAvailablaSnapshots] = useState<string[]>([]);
  const [dates, setDates] = useState<SortedDates | undefined>();

  useEffect(() => {
    const formatDates = () => {
      const allDates = uniq([today, ...props.snapshots]);
      setAvailablaSnapshots(allDates);
      setDates(groupDatesByYearAndMonth(allDates));
    };
    formatDates();
  }, [props.snapshots]); /* eslint-disable-line react-hooks/exhaustive-deps */

  if (isUndefined(dates) || availablaSnapshots.length < MIN_NUMBER_SNAPSHOTS) return null;

  return (
    <div className="d-none d-sm-flex my-4 my-md-5 ms-4 ms-md-5">
      <div>
        <div className="d-flex flex-column text-center border">
          <div className="mt-2 pb-3 fs-4">
            <RiHistoryFill />
          </div>
          <div className="d-flex flex-column text-center">
            {Object.keys(dates)
              .sort()
              .reverse()
              .map((year: string) => {
                return (
                  <div className="mb-3" key={`year_${year}`}>
                    <div className={`fw-bold w-100 border-top border-bottom px-1 mb-3 ${styles.year}`}>{year}</div>
                    {Object.keys(dates[year])
                      .sort()
                      .reverse()
                      .map((month: string) => {
                        return (
                          <Fragment key={`date_${year}_${month}`}>
                            <div className={`position-relative mt-3 mb-2 fw-bold text-uppercase ${styles.month}`}>
                              {moment.monthsShort(parseInt(month) - 1)}
                            </div>
                            <div className="d-flex flex-column align-items-center">
                              {dates[year][month]
                                .sort()
                                .reverse()
                                .map((time: string) => {
                                  const isToday = time === today;
                                  const selectedDate = props.activeDate || today;
                                  const isActive = selectedDate === time;

                                  return (
                                    <button
                                      key={`time_${time}`}
                                      className={classNames(
                                        'position-relative btn btn-link text-decoration-none text-center text-muted rounded-circle my-1 p-0',
                                        styles.dot,
                                        {
                                          [styles.activeDot]: isActive,
                                        }
                                      )}
                                      onClick={() => {
                                        props.setActiveDate(!isToday ? time : undefined);
                                      }}
                                      disabled={isActive}
                                      aria-label={`Opens snapshot: ${moment(time, 'YYYY-MM-DD').format("Do MMM 'YY")}`}
                                    >
                                      <div className={styles.content}>{moment(time, 'YYYY-MM-DD').format('D')}</div>
                                    </button>
                                  );
                                })}
                            </div>
                          </Fragment>
                        );
                      })}
                  </div>
                );
              })}
          </div>
        </div>
      </div>
    </div>
  );
};

export default Timeline;
