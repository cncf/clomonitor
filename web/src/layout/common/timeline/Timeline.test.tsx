import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Timeline from './Timeline';

const mockSetActiveDate = jest.fn();

const defaultProps = {
  snapshots: ['2022-10-28', '2022-10-27', '2022-10-26', '2022-09-10', '2022-08-07', '2021-04-18', '2020-03-09'],
  setActiveDate: mockSetActiveDate,
};

describe('Timeline', () => {
  let dateNowSpy: any;

  beforeEach(() => {
    dateNowSpy = jest.spyOn(Date, 'now').mockImplementation(() => 1667029023000);
  });

  afterAll(() => {
    dateNowSpy.mockRestore();
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(<Timeline {...defaultProps} />);

    expect(asFragment()).toMatchSnapshot();
  });

  describe('Render', () => {
    it('renders timeline', async () => {
      render(<Timeline {...defaultProps} />);

      const buttons = screen.getAllByRole('button', { name: /Opens snapshot:/ });
      expect(buttons).toHaveLength(8);
      expect(buttons[0]).toHaveTextContent('29');
      expect(buttons[1]).toHaveTextContent('28');
      expect(buttons[2]).toHaveTextContent('27');
      expect(buttons[3]).toHaveTextContent('26');
      expect(buttons[4]).toHaveTextContent('10');
      expect(buttons[5]).toHaveTextContent('7');
      expect(buttons[6]).toHaveTextContent('18');
      expect(buttons[7]).toHaveTextContent('9');

      expect(screen.getByText('2022')).toBeInTheDocument();
      expect(screen.getByText('2021')).toBeInTheDocument();
      expect(screen.getByText('2020')).toBeInTheDocument();
      expect(screen.getByText('Oct')).toBeInTheDocument();
      expect(screen.getByText('Sep')).toBeInTheDocument();
      expect(screen.getByText('Aug')).toBeInTheDocument();
      expect(screen.getByText('Apr')).toBeInTheDocument();
      expect(screen.getByText('Mar')).toBeInTheDocument();
    });

    it('clicks day button', async () => {
      const { rerender } = render(<Timeline {...defaultProps} />);

      const buttons = screen.getAllByRole('button', { name: /Opens snapshot:/ });
      expect(buttons[0]).toHaveClass('activeDot');

      await userEvent.click(buttons[3]);

      await waitFor(() => {
        expect(mockSetActiveDate).toHaveBeenCalledTimes(1);
        expect(mockSetActiveDate).toHaveBeenCalledWith('2022-10-26');

        rerender(<Timeline {...defaultProps} activeDate="2022-10-26" />);

        expect(buttons[3]).toHaveClass('activeDot');
      });
    });

    it('clicks Now day button', async () => {
      render(<Timeline {...defaultProps} activeDate="2022-10-26" />);

      const buttons = screen.getAllByRole('button', { name: /Opens snapshot:/ });
      expect(buttons[3]).toHaveClass('activeDot');

      await userEvent.click(buttons[0]);

      await waitFor(() => {
        expect(mockSetActiveDate).toHaveBeenCalledTimes(1);
        expect(mockSetActiveDate).toHaveBeenCalledWith(undefined);
      });
    });

    describe('does not render', () => {
      it('when snapshots list is empty', () => {
        const { container } = render(<Timeline {...defaultProps} snapshots={[]} />);

        expect(container).toBeEmptyDOMElement();
      });

      it('when snapshots list + today length is less than 3 items', () => {
        const { container } = render(<Timeline {...defaultProps} snapshots={['2022-10-28']} />);

        expect(container).toBeEmptyDOMElement();
      });
    });
  });
});
