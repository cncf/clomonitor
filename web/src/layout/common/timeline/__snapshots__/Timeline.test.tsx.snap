// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="d-none d-sm-flex undefined"
  >
    <div>
      <div
        class="d-flex flex-column text-center border"
      >
        <div
          class="mt-2 pb-3 fs-4"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g>
              <path
                d="M0 0H24V24H0z"
                fill="none"
              />
              <path
                d="M12 2c5.523 0 10 4.477 10 10s-4.477 10-10 10S2 17.523 2 12h2c0 4.418 3.582 8 8 8s8-3.582 8-8-3.582-8-8-8C9.536 4 7.332 5.114 5.865 6.865L8 9H2V3l2.447 2.446C6.28 3.336 8.984 2 12 2zm1 5v4.585l3.243 3.243-1.415 1.415L11 12.413V7h2z"
              />
            </g>
          </svg>
        </div>
        <div
          class="d-flex flex-column text-center"
        >
          <div
            class="mb-3"
          >
            <div
              class="fw-bold w-100 border-top border-bottom px-1 mb-3 year"
            >
              2022
            </div>
            <div
              class="position-relative mt-3 mb-2 fw-bold text-uppercase month"
            >
              Oct
            </div>
            <div
              class="d-flex flex-column align-items-center"
            >
              <button
                aria-label="Opens snapshot: 29th Oct '22"
                class="position-relative btn btn-link text-decoration-none text-center text-muted rounded-circle my-1 p-0 dot activeDot"
                disabled=""
              >
                <div
                  class="content"
                >
                  29
                </div>
              </button>
              <button
                aria-label="Opens snapshot: 28th Oct '22"
                class="position-relative btn btn-link text-decoration-none text-center text-muted rounded-circle my-1 p-0 dot"
              >
                <div
                  class="content"
                >
                  28
                </div>
              </button>
              <button
                aria-label="Opens snapshot: 27th Oct '22"
                class="position-relative btn btn-link text-decoration-none text-center text-muted rounded-circle my-1 p-0 dot"
              >
                <div
                  class="content"
                >
                  27
                </div>
              </button>
              <button
                aria-label="Opens snapshot: 26th Oct '22"
                class="position-relative btn btn-link text-decoration-none text-center text-muted rounded-circle my-1 p-0 dot"
              >
                <div
                  class="content"
                >
                  26
                </div>
              </button>
            </div>
            <div
              class="position-relative mt-3 mb-2 fw-bold text-uppercase month"
            >
              Sep
            </div>
            <div
              class="d-flex flex-column align-items-center"
            >
              <button
                aria-label="Opens snapshot: 10th Sep '22"
                class="position-relative btn btn-link text-decoration-none text-center text-muted rounded-circle my-1 p-0 dot"
              >
                <div
                  class="content"
                >
                  10
                </div>
              </button>
            </div>
            <div
              class="position-relative mt-3 mb-2 fw-bold text-uppercase month"
            >
              Aug
            </div>
            <div
              class="d-flex flex-column align-items-center"
            >
              <button
                aria-label="Opens snapshot: 7th Aug '22"
                class="position-relative btn btn-link text-decoration-none text-center text-muted rounded-circle my-1 p-0 dot"
              >
                <div
                  class="content"
                >
                  7
                </div>
              </button>
            </div>
          </div>
          <div
            class="mb-3"
          >
            <div
              class="fw-bold w-100 border-top border-bottom px-1 mb-3 year"
            >
              2021
            </div>
            <div
              class="position-relative mt-3 mb-2 fw-bold text-uppercase month"
            >
              Apr
            </div>
            <div
              class="d-flex flex-column align-items-center"
            >
              <button
                aria-label="Opens snapshot: 18th Apr '21"
                class="position-relative btn btn-link text-decoration-none text-center text-muted rounded-circle my-1 p-0 dot"
              >
                <div
                  class="content"
                >
                  18
                </div>
              </button>
            </div>
          </div>
          <div
            class="mb-3"
          >
            <div
              class="fw-bold w-100 border-top border-bottom px-1 mb-3 year"
            >
              2020
            </div>
            <div
              class="position-relative mt-3 mb-2 fw-bold text-uppercase month"
            >
              Mar
            </div>
            <div
              class="d-flex flex-column align-items-center"
            >
              <button
                aria-label="Opens snapshot: 9th Mar '20"
                class="position-relative btn btn-link text-decoration-none text-center text-muted rounded-circle my-1 p-0 dot"
              >
                <div
                  class="content"
                >
                  9
                </div>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
